src/modules/extrakeys/hsort_impl.h

```diff
@@ -0,0 +1,104 @@
+/***********************************************************************
+ * Copyright (c) 2021 Russell O'Connor, Jonas Nick                     *
+ * Distributed under the MIT software license, see the accompanying    *
+ * file COPYING or https://www.opensource.org/licenses/mit-license.php.*
+ ***********************************************************************/
+
+#ifndef SECP256K1_HSORT_IMPL_H_
+#define SECP256K1_HSORT_IMPL_H_
+
+#include "hsort.h"
+
+/* An array is a heap when, for all non-zero indexes i, the element at index i
+ * compares as less than or equal to the element at index parent(i) = (i-1)/2.
+ */
+
+static SECP256K1_INLINE size_t child1(size_t i) {
+    VERIFY_CHECK(i <= (SIZE_MAX - 1)/2);
+    return 2*i + 1;
+}
+
+static SECP256K1_INLINE size_t child2(size_t i) {
+    VERIFY_CHECK(i <= SIZE_MAX/2 - 1);
+    return child1(i)+1;
+}
+
+static SECP256K1_INLINE void swap64(unsigned char *a, size_t i, size_t j, size_t stride) {
+    unsigned char tmp[64];
+    VERIFY_CHECK(stride <= 64);
+    memcpy(tmp, a + i*stride, stride);
+    memmove(a + i*stride, a + j*stride, stride);
+    memcpy(a + j*stride, tmp, stride);
+}
+
+static SECP256K1_INLINE void swap(unsigned char *a, size_t i, size_t j, size_t stride) {
+    while (64 < stride) {
+        swap64(a + (stride - 64), i, j, 64);
+        stride -= 64;
+    }
+    swap64(a, i, j, stride);
+}
+
+static SECP256K1_INLINE void heap_down(unsigned char *a, size_t i, size_t heap_size, size_t stride,
+                            int (*cmp)(const void *, const void *, void *), void *cmp_data) {
+    while (i < heap_size/2) {
+        VERIFY_CHECK(child1(i) < heap_size);
+        /* Proof:
+         * i < heap_size/2
+         * i + 1 <= heap_size/2
+         * 2*i + 2 <= heap_size
+         * 2*i + 1 < heap_size
+         * child1(i) < heap_size
+         */
+
+        VERIFY_CHECK(i <= SIZE_MAX/2 - 1);
+        /* Proof:
+         * i < heap_size/2
+         * i + 1 <= heap_size/2
+         * 2*i + 2 <= heap_size <= SIZE_MAX
+         * 2*i <= SIZE_MAX - 2
+         */
+
+        /* If any child of i is larger than i, swap the i with that child. If
+         * both children of i are larger than i, swap i with the larger child to
+         * ensure the new parent is larger than both children.
+         */
+        if (child2(i) < heap_size && 0 < cmp(a + child2(i)*stride, a + child1(i)*stride, cmp_data)
+                              && 0 < cmp(a + child2(i)*stride, a +         i*stride, cmp_data)) {
```

Comment:
It might be slightly better to break the comparison of child2 with i into its own if statement, because if this comparison fails (after having checked that child2 is greater than (or equal to) child1) then it must be the case that, by transitivity, that i is also greater than or equal to child1, and thus the next comparison on the line below is going to fail, and we can skip it.

Thus my suggestion is to replace this branch with:

    if (0 < cmp(a + child2(i)*stride, a +         i*stride, cmp_data)) {
        swap(a, i, child2(i), stride);
        i = child2(i);
    } else {
        return;
    }

Reply:
Done, also added comment explaining this.